### ISSUE TITLE
swig 4.2.0 generates wrong C# code

### ISSUE THREAD

**Issue by paroj** [swig/swig#2829]
for this strongly typed enum:
https://github.com/OGRECave/ogre/blob/3f3a4be1b0e004d9ffd0e8188d8d8a035a114df6/OgreMain/include/OgreFrustum.h#L47

with 4.2.0/ 4.2.1 I get:
```csharp
public enum ProjectionType : $&csclassname {
  PT_ORTHOGRAPHIC,
  PT_PERSPECTIVE
}
```
I think the `$&csclassname` is wrongly emitted into the source code.

with 4.1.1 I get:
```csharp
public enum ProjectionType {
  PT_ORTHOGRAPHIC,
  PT_PERSPECTIVE
}
```

**Comment by paroj** [comment]
ok, basically it is my fault. `typedef uint8_t uint8;` was missing. However, I think the error reporting could be improved here.

**Comment by ojwb** [comment]
This seems like a bug to me - SWIG is meant to be able to work with incomplete type information, and there's a reasonable thing to do when it doesn't know the C++ base type of the enum - just omit the base type in C#, like 4.1.1 does.

Here's a self-contained reproducer:

```
%module x
%{
typedef unsigned char uint8;
%}
%inline %{
enum E : uint8 { E1, E2 };
%}
```

This should produce no output:

```
swig -c++ -csharp test.i && grep csclassname *.cs
```

`git bisect` shows the first bad commit as af72fbb08b23d6922ba5cd1d44020376a26b4714 which is very plausible - @wsfulton that's one of yours.

**Comment by wsfulton** [comment]
There is a missing expansion of the `$csclassname` family of variables into proxy class names/type wrapper classes in this particular case. A C# proxy for an enum base would be unusual, or possibly even not possible. Type wrappers almost always indicate missing type information so, I've fixed this missing expansion and improved the docs for handling missing type information for wrapping C++ base types.

**Comment by wsfulton** [comment]
Closed by bc0b8b558de47388de30e7f9aa689fc7049d41f7.
